### PR TITLE
SUBMARINE-701. Support Tensorboard in Experiment

### PR DESCRIPTION
**Pull request by ByronHsu**
### What is this PR for?

Support new feature in 0.6.0: tensorboard integration.

- Usage
    1. Create a job request that uses tensorboard
    2. Write tensorboard log to `/logs/mylog` (The subpath is required due to this issue in tensorflow [https://github.com/kubeflow/tf-operator/issues/1053](https://github.com/kubeflow/tf-operator/issues/1053). We cannot directly write log file to mountPath.)
    3. Link to `http://<host>:<ip>/tfboard-${job-name}`, and you can monitor the tensorboard with ease!
- Implementation

    When creating a new job, the backend will not only create original experiment but also several k8s resources required in tensorboard

    The resources can be classified into two categories: 

    1. Storage
    2. Tensorboard serving

    **Storage**

    The resources required for storage are **persistent volume** and **persistent volume claim**.  

    I set the storage path of persistent volume on host path, and mount this path to MLjob (enable job to generate logs to volume) and Tensorboard (enable tfboard to access logs).

    **Tensorboard Serving**

    The resources required here are **deployments, service, and ingressroute**.

    I create the tensorboard apps with deployments and service, and then redirect it to custom path with the help of ingressroute.

- Example
    - tensorboard-example.json

        ```bash
        {
          "meta": {
            "name": "tensorflow-dist-mnist-byron-1234",
            "namespace": "default",
            "framework": "TensorFlow",
            "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/logs/mylog --learning_rate=0.01 --batch_size=20",
            "envVars": {
              "ENV_1": "ENV1"
            }
          },
          "environment": {
            "image": "apache/submarine:tf-mnist-with-summaries-1.0"
          },
          "spec": {
            "Worker": {
              "replicas": 1,
              "resources": "cpu=1,memory=1024M"
            }
          }
        }
        ```
 
        Kapture 2020-12-27 at 18 04 40

### What type of PR is it?
[Feature]

### Todos
- [ ] Frontend support
- [ ] The logs of job cannot be written directly on the mountPath (As describe in above). We should fix this problem.
- [ ] Make log path configurable (Currently, it is hard-coded as `/logs` )
- [ ] Support smb-server for shared storage

### What is the Jira issue?
https://issues.apache.org/jira/projects/SUBMARINE/issues/SUBMARINE-701

### How should this be tested?
https://travis-ci.org/github/ByronHsu/submarine/jobs/751658488

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: helm-charts/submarine/templates/rbac.yaml [hunk 1]
```diff
@@ -56,6 +56,14 @@ rules:
...
   - pods
   - pods/log
+  - services
+  - persistentvolumes
+  - persistentvolumeclaims
+  verbs:
+  - '*'
+- apiGroups: ["extensions", "apps"]
+  resources:
+  - deployments
```
Comment (lowc1012) [inline helm-charts/submarine/templates/rbac.yaml:66]: The Deployment is in API version "apps/v1"  after k8s v.1.9

**Comment by ByronHsu** [comment]
@lowc1012 @jiwq @xunliu @pingsutw please help me review the code. Thanks!

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/experiment/ExperimentManager.java [hunk 2]
```diff
@@ -101,6 +101,9 @@ public Experiment createExperiment(ExperimentSpec spec) throws SubmarineRuntimeE
 
     Experiment experiment = submitter.createExperiment(spec);
     experiment.setExperimentId(id);
+    // importing tensorboardUtils will cause dependency circle. Hard-code it as a temporary solution
```
Comment (pingsutw) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/experiment/ExperimentManager.java:104]:
```suggestion
    // [TODO] Importing tensorboardUtils will cause a dependency circle. Hard-code it as a temporary solution
```

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/ExperimentSpecParser.java [hunk 3]
```diff
@@ -174,7 +176,18 @@ private static V1PodTemplateSpec parseTemplateSpec(
     resources.setLimits(parseResources(taskSpec));
     container.setResources(resources);
     container.setEnv(parseEnvVars(taskSpec, experimentSpec.getMeta().getEnvVars()));
-    
+
+    // volumeMount
+    container.addVolumeMountsItem(new V1VolumeMount().mountPath("/logs").name("volume"));
+
+    // volume
+    final String PVC_NAME_PREFIX = "tfboard-pvc-";
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/ExperimentSpecParser.java:184]: We could use `PVC_NAME_PREFIX` in `TensorboardUtils`

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/TensorboardSpecParser.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    return deployment;
+  }
+
+  public static V1Service parseService(String svc_name, String pod_name) {
+    V1Service svc = new V1Service();
+    svc.metadata(new V1ObjectMeta().name(svc_name));
+
+    V1ServiceSpec svc_spec = new V1ServiceSpec();
+    svc_spec.setSelector(Collections.singletonMap("app", pod_name)); // bind to pod
+    svc_spec.addPortsItem(new V1ServicePort().protocol("TCP").targetPort(new IntOrString(6006)).port(8080));
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/TensorboardSpecParser.java:115]: You could create constant variables in `TensorboardUtils`  
`final String DEFAULT_TENSORBOARD_PORT = "6006";`
`final String EXPOSE_TENSORBOARD_PORT= "8080";`

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/TensorboardSpecParser.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    IngressRouteSpec ingressRoute_spec = new IngressRouteSpec();
+    ingressRoute_spec.setEntryPoints(new HashSet<>(Collections.singletonList("web")));
+    SpecRoute spec_route = new SpecRoute();
+    spec_route.setKind("Rule");
+    spec_route.setMatch(String.format("PathPrefix(`%s`)", route_path));
+
+    Map<String, Object> service = new HashMap<String, Object>() {{
+        put("kind", "Service");
+        put("name", svc_name);
+        put("port", 8080);
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/TensorboardSpecParser.java:137]:
```suggestion
        put("port", EXPOSE_TENSORBOARD_PORT);
```

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/TensorboardSpecParser.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+
+public class TensorboardSpecParser {
+  public static V1Deployment parseDeployment(String name, String image, String route_path, String pvc) {
+
+    /*
+    [start] Deployment
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/TensorboardSpecParser.java:51]: The comments here seem unnecessary, could remove them?
This code can be easily understood

File: submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/K8SJobSubmitterTest.java [hunk 7]
```diff
@@ -70,7 +72,22 @@ public void testRunTFJobPerRequest() throws URISyntaxException,
...
+    ExperimentSpec spec = (ExperimentSpec) buildFromJsonFile(ExperimentSpec.class, tfTfboardJobwReqFile);
+    Experiment experiment = submitter.createExperiment(spec);
+    Assert.assertTrue(true);
+  }
+
+  @Test
+  public void testDeleteTFJob() throws IOException, URISyntaxException {
+    ExperimentSpec spec = (ExperimentSpec) buildFromJsonFile(ExperimentSpec.class, tfTfboardJobwReqFile);
+    Experiment experiment = submitter.deleteExperiment(spec);
+    Assert.assertTrue(true);
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/K8SJobSubmitterTest.java:86]:
```suggestion
```

File: submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/K8SJobSubmitterTest.java [hunk 8]
```diff
@@ -70,7 +72,22 @@ public void testRunTFJobPerRequest() throws URISyntaxException,
     run(spec);
   }
 
+  @Test
+  public void testCreateTFJob() throws IOException, URISyntaxException {
+    ExperimentSpec spec = (ExperimentSpec) buildFromJsonFile(ExperimentSpec.class, tfTfboardJobwReqFile);
+    Experiment experiment = submitter.createExperiment(spec);
+    Assert.assertTrue(true);
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/K8SJobSubmitterTest.java:79]:
```suggestion
```

File: submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/K8SJobSubmitterTest.java [hunk 9]
```diff
@@ -70,7 +72,22 @@ public void testRunTFJobPerRequest() throws URISyntaxException,
...
+
+  @Test
+  public void testDeleteTFJob() throws IOException, URISyntaxException {
+    ExperimentSpec spec = (ExperimentSpec) buildFromJsonFile(ExperimentSpec.class, tfTfboardJobwReqFile);
+    Experiment experiment = submitter.deleteExperiment(spec);
+    Assert.assertTrue(true);
+  }
+
   private void run(ExperimentSpec spec) throws SubmarineRuntimeException {
+    System.out.println(spec.toString());
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/K8SJobSubmitterTest.java:90]:
```suggestion
```

File: submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/parser/VolumeSpecParserTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class VolumeSpecParserTest  {
+  private static final Logger LOG = LoggerFactory.getLogger(VolumeSpecParserTest.class);
+
+  @Before
+  public void before() {
+
```
Comment (pingsutw) [inline submarine-server/server-submitter/submitter-k8s/src/test/java/org/apache/submarine/server/submitter/k8s/parser/VolumeSpecParserTest.java:36]:
```suggestion
```

**Review by pingsutw — COMMENTED** [review]
@ByronHsu Thank you for your great work
I have made Some suggestions below.

**Review by jiwq — APPROVED** [review]
+1 Thanks @ByronHsu for the great works and @pingsutw for the review.
Will merge if no more comments.

**Pull request closed without merging**